File: 算法/代码随想录算法PDF/00调试源代码/17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#

# @lc code=start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        s = ""
        letterMap = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        def backtracking(index):
            nonlocal s
            if index == len(digits):
                res.append(s)
                return
            letters = letterMap[digits[index]]
            for letter in letters:
                s += letter
                backtracking(index + 1)
                s = s[:-1]
        
        if len(digits) == 0:
            return res
        backtracking(0)
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/17.电话号码的字母组合.cpp
/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
private:
    const string letterMap[10] = {
        "",
        "",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };
public:
    vector<string> res;
    string s;
    void backtracking(string& digits, int index) {
        if (index == digits.size()) {
            res.push_back(s);
            return ;
        }
        string letters = letterMap[digits[index] - '0'];  // 取到此时的字符集
        for (char c : letters) {
            s.push_back(c);
            backtracking(digits, index + 1);
            s.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) return res;
        backtracking(digits, 0);
        return res;
    }
};
// @lc code=end

